crates/burn-autodiff/src/runtime/memory_management.rs

```diff
@@ -84,7 +84,10 @@ impl GraphMemoryManagement {
         for (id, parents) in self.nodes.iter() {
             let is_useful = matches!(self.statuses.get(id), Some(NodeMemoryStatus::Useful));
 
-            if !is_useful && Arc::strong_count(id) == 1 && parents.is_empty() {
+            let parents_absent =
+                parents.is_empty() || parents.iter().all(|p| !self.nodes.contains_key(p));
```

Comment:
If `parents.is_empty() == true`, `parents.iter().all(...) == true`. If the second expression works, the first expression i.e. `is_empty` has no effect.

https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.all